### PR TITLE
feat: Python class meta-programming for State construction

### PR DESCRIPTION
**Pull request by kyle-cochran**
This MR is some Python-only syntactic sugar on top of the State Builder concept which allows for some nice syntax:
```python
MySuperFunState: type = State.template(
    frame,
    [CartesianPosition.default(), CartesianVelocity.default()],
)

states: MySuperFunState = [
    MySuperFunState(instant,coords) for instant,coords in zip(instants, coordinates)
]
```

The output StateTypeX classes are polymorphic to State and so can be used anywhere a State can.
This example is 100% equivalent to using a builder and is just a syntax alternative.

(other meta-programming features like auto-assigning `state.mass` to get the mass value will be in a follow-up)

### CODE REVIEW

File: bindings/python/test/trajectory/test_state.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from ostk.physics.coordinate import Velocity
 from ostk.physics.coordinate import Frame
 
-from ostk.astrodynamics.trajectory import State
+from ostk.astrodynamics.trajectory import State, StateBuilder
```
Comment (kyle-cochran) [inline bindings/python/test/trajectory/test_state.py:14]:
```suggestion
from ostk.astrodynamics.trajectory import State
```

File: bindings/python/tools/python/ostk/astrodynamics/__init__.py [hunk 2]
```diff
@@ -3,3 +3,9 @@
 from ostk.physics import *
 
 from .OpenSpaceToolkitAstrodynamicsPy import *
+
+from .pytrajectory.pystate import State as PyState
```
Comment (kyle-cochran) [inline bindings/python/tools/python/ostk/astrodynamics/__init__.py:7]: Curious if there are any better ideas here. This works, but it means State is imported whenever anything else in Astro is imported. State is already very pervasive, so it may not make much difference, but it's still messy.

File: bindings/python/test/trajectory/test_state.py [hunk 3]
```diff
@@ -105,6 +105,38 @@ def test_subsets_constructor(
         assert isinstance(state, State)
         assert state.is_defined()
 
+    def test_custom_class_generator(
```
Comment (vishwa2710) [inline bindings/python/test/trajectory/test_state.py:108]:
```suggestion
    def test_custom_state_class_generator(
```
Nit af:

File: include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp [hunk 4]
```diff
@@ -83,6 +83,19 @@ class State
...
 
+    /// @brief                  Copy constructor.
+    ///
+    /// @param                  [in] aState An existing State
+
+    State(const State& aState);
+
+    /// @brief                  Copy-assignment operator
+    ///
+    /// @param                  [in] aState The State to compare to
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp:94]:
```suggestion
    /// @param                  [in] aState The State to copy
```

File: include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp [hunk 5]
```diff
@@ -83,6 +83,19 @@ class State
 
     State(const Instant& anInstant, const Position& aPosition, const Velocity& aVelocity);
 
+    /// @brief                  Copy constructor.
+    ///
+    /// @param                  [in] aState An existing State
+
+    State(const State& aState);
+
+    /// @brief                  Copy-assignment operator
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp:92]: Why are we adding this? 🤔 Doesn't seem related to the Meta-class?

**Review by vishwa2710 — APPROVED** [review]
LGTM!
It's gonna be sick when we add the custom kwargs and getters 🥳

File: include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp (hunk 5 above)
Comment (kyle-cochran) [reply include/OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp:92]: The copy constructor is needed here:
https://github.com/open-space-collective/open-space-toolkit-astrodynamics/pull/244/files#diff-f995720c611c0d79a20e7b908e36d9360719bb3784955f510982c459db799006R32

When sub-classing a `State`, we need to call it's `__init__` during construction of the sub-class. In this case, I don't want the `State` object to actually be produced using the normal `State` constructor, but via a `StateBuilder`.

There could be a way around this by ignoring the `StateBuilder`, but I think it's a good idea to use it so we never have to re-code all of the extra stuff inside of it.


The copy assignment operator is because after creating the copy constructor, the compiler was giving warnings that it could not use the default copy operator anymore, so I added an explicit one.

**Review by phc1990 — COMMENTED** [review]
I wonder whether we can achieve this by using bindings directly and not Python code 🤔 
I just dont' know
